src/ui/public/state_management/state_storage/hashed_item_store.js

```diff
@@ -0,0 +1,165 @@
+ *
+ * Old < - - - - - - - - > New
+ * A1 | B1 | C1 | A2 | D1 | E1
+ *
+ * If the user navigates back to C1 and starts to create new states, persisted history states will
+ * become inaccessible:
+ *
+ * Old < - - - - - - - - - - -> New
+ * A1 | B1 | C1 | F1 | G1 | H1 | I1  (new history states)
+ *                A2 | D1 | E1       (inaccessible persisted history states)
+ *
+ * Theoretically, we could build a mirror of the browser history. When the onpopstate event is
+ * dispatched, we could determine whether we have gone back or forward in history. Then, when
+ * a new state is persisted, we could delete all of the persisted items which are no longer
+ * accessible. (Note that this would require reference-counting so that A isn't removed while D and
+ * E are, since A would still have a remaining reference from A1).
+ *
+ * However, the History API doesn't allow us to read from the history beyond the current state. This
+ * means that if a session is restored, we can't rebuild this browser history mirror.
+ *
+ * Due to this imperfect implementation, HashedItemStore ignores the possibility of inaccessible
+ * history states. In the future, we could implement this history mirror and persist it in
+ * sessionStorage too. Then, when restoring a session, we can just retrieve it from sessionStorage.
+ *
+ * -------------------------------------------------------------------------------------------------
+ *
+ * Consideration 2: We can't tell when we've hit the browser history limit
+ *
+ * Because some of our persisted history states may no longer be referenced by the browser history,
+ * and we have no way of knowing which ones, we have no way of knowing whether we've persisted a
+ * number of accessible states beyond the browser history length limit.
+ *
+ * More fundamentally, the browser history length limit is a browser implementation detail, so it
+ * can change from browser to browser, or over time. Respecting this limit would introduce a lot of
+ * (unnecessary?) complexity.
+ *
+ * For these reasons, HashedItemStore doesn't concern itself with this constraint.
+ */
+
+export default class HashedItemStore {
+
+  /**
+   * HashedItemStore uses objects called indexed items to refer to items that have been persisted
+   * in sessionStorage. An indexed item is shaped {hash, touched}. The touched date is when the item
+   * was last referenced by the browser history.
+   */
+  constructor(sessionStorage = window.sessionStorage) {
+    this._sessionStorage = sessionStorage;
+
+    // Store indexed items in descending order by touched (oldest first, newest last). We'll use
+    // this to remove older items when we run out of storage space.
+    this._items = [];
+    // Associate item hashes with the corresponding indexed items. We'll use this to quickly look
+    // up an item and update its touched date when it reoccurs in the browser history.
+    this._itemIndex = {};
```

Comment:
I thought of that too but decided it wasn't appropriate because the thing that's being stored isn't actually an `item` (but an indexing-specific object with shape `{hash, touched}`). How about `_hashToIndexedItemMap`? Verbose but very accurate.